pkg/api/heartbeat.go

```diff
@@ -25,55 +26,139 @@ func (c *Client) SendHeartbeats(heartbeats []heartbeat.Heartbeat) ([]heartbeat.R
 
 	log.Debugf("sending %d heartbeat(s) to api at %s", len(heartbeats), url)
 
+	grouped := groupByApiKey(heartbeats)
+
+	cherr := make(chan error, len(grouped))
+	defer close(cherr)
+
+	chres := make(chan []heartbeat.Result, len(grouped))
+	defer close(chres)
+
+	// don't spawn threads when there's only one api key set.
+	if len(grouped) == 1 {
+		c.sendHeartbeats(url, heartbeats, chres, cherr)
+
+		return <-chres, <-cherr
+	}
+
+	var wg sync.WaitGroup
+
+	for apiKey, hh := range grouped {
+		hh := hh
```

Comment:
Why do we re-define these?

Reply:
https://go.dev/doc/faq#closures_and_goroutines

Reply:
Ah, same as JavaScript. Thanks!